Dockerfile

```diff
@@ -1,22 +1,45 @@
-FROM python:3.7-slim-stretch
+FROM node:current-alpine AS NODEBUILD
+# copy directory to /opt and work inside it
+WORKDIR /opt
+COPY . /opt
+# run the npm portion of build-static script
+RUN rm -rf static && mkdir -p static/
+RUN npm install -g npm && \
+    npm install && \
+    npm run build
 
-# Install packages required for uswgi
-RUN apt-get update
-RUN apt-get -y install build-essential
-RUN apt-get -y install mime-support
 
-# Install requirements and uwsgi server for running python web apps
+FROM python:3.9-slim AS PYTHONBUILD
+# install missing system deps
+RUN apt-get update && apt-get -y install build-essential mime-support libpcre3-dev libsass-dev mime-support
+# set working dir
 WORKDIR /etc/linkding
 COPY requirements.prod.txt ./requirements.txt
-RUN pip install -U pip
-RUN pip install -Ur requirements.txt
+# make a virtualenv, upgrade pip & wheel, install prod requirements
+RUN mkdir /venv && \
+    python -m venv --upgrade-deps --copies /venv && \
+    /venv/bin/pip install --upgrade pip wheel && \
+    /venv/bin/pip install -Ur requirements.txt
+# run the python half of build-static script
+COPY --from=NODEBUILD /opt/ .
+RUN /venv/bin/python manage.py compilescss && \
+    /venv/bin/python manage.py collectstatic --ignore=*.scss && \
+    /venv/bin/python manage.py compilescss --delete-files
 
-# Copy application
-COPY . .
 
-# Expose uwsgi server at port 9090
+FROM python:3.9-slim as FINAL
+WORKDIR /etc/linkding
+# copy application code from PYTHONBUILD stage
+COPY --from=PYTHONBUILD /etc/linkding  /etc/linkding
+# copy virtualenv from PYTHONBUILD stage
+COPY --from=PYTHONBUILD /venv /venv
+# copy mime.types from PYTHONBUILD stage
+COPY --from=PYTHONBUILD /etc/mime.types /etc/mime.types
```

Comment:
Not sure about this. Would it be a big deal to run this as an apt-get in this stage? It should be a new layer in any case, right? Plus, I think it's not needed in the build stage.

Also there is a diff here in that `build-essentials` is missing. Not sure why that was required before - maybe because of the `pip install`? Do you have any insights here?

Reply:
RE: build-essentials - apparently required for installing uwsgi.

Reply:
This is all personal preference. I always package into virtualenvs to separate concerns (and having better reproducability between dev and prod). And yeah, `build-essential` is for compiling uwsgi when installed via pip.